File: clients/haskell/src/Debugging/Camera.hs
module Debugging.Camera where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Model.Vec2 (Vec2)

-- | Camera state
data Camera = Camera {
    -- | Center
    center :: Vec2,
    -- | Rotation
    rotation :: Double,
    -- | Attack angle
    attack :: Double,
    -- | Vertical field of view
    fov :: Double }
    deriving Show

instance Trans Camera where
    read = do
        center <- Trans.read
        rotation <- Trans.read
        attack <- Trans.read
        fov <- Trans.read
        return Camera {
            center,
            rotation,
            attack,
            fov }
    
    write Camera {
        center,
        rotation,
        attack,
        fov } = do
            Trans.write center
            Trans.write rotation
            Trans.write attack
            Trans.write fov

File: clients/rust/src/debugging/camera.rs
use super::*;

/// Camera state
#[derive(Clone, Debug)]
pub struct Camera {
    /// Center
    pub center: model::Vec2,
    /// Rotation
    pub rotation: f64,
    /// Attack angle
    pub attack: f64,
    /// Vertical field of view
    pub fov: f64,
}

impl trans::Trans for Camera {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.center.write_to(writer)?;
        self.rotation.write_to(writer)?;
        self.attack.write_to(writer)?;
        self.fov.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let center: model::Vec2 = trans::Trans::read_from(reader)?;
        let rotation: f64 = trans::Trans::read_from(reader)?;
        let attack: f64 = trans::Trans::read_from(reader)?;
        let fov: f64 = trans::Trans::read_from(reader)?;
        Ok(Self {
            center,
            rotation,
            attack,
            fov,
        })
    }
}